front: add support for importing NGE JSON files

emersion opened the pull request:
Test plan:

- Open https://nge.flatland.cloud
- Create a project there, add a bunch of trainruns
- Export as JSON
- Import in OSRD

Closes: https://github.com/osrd-project/osrd-confidential/issues/1081

emersion committed:
front: avoid implicit any cast in processJsonFile()

JSON.parse() returns "any", which can be silently cast to any other
type. Instead, use "unknown" which requires an explicit cast, to
make it clear what's going on and that the parsed file can have an
arbitrary JSON payload.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: unify paced field in ngeToOsrd

There are two spots where we need to create the paced field: when
a new trainrun is created, and when a trainrun is updated. We used
to have two different codepaths for each. Refactor
createPacedAttributesFromTrainrun() so that it can be used from
handleUpdateTimetableItem().

The null check operator "!" makes it explicit in the code that we
assume a new trainrun will never have TRAIN_SCHEDULE_FREQUENCY_ID
(see the comment above handleCreateTimetableItem()).

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: move exceptions field to DEFAULT_PACED_TRAIN_PAYLOAD

That way we don't need to repeat it everywhere this const is used.
(It's only used once for now, but will be used more in the next
commits.)

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: remove check for empty string in secondary code

We believed at some point that we had OPs with an empty string as
seconary code to indicate a station. However this turned out not
to be the case, but this piece of code was left behind.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: split station OP search into separate function in ngeToOsrd

When importing a full NGE DTO, we don't want to query the API to
find the secondary code of OPs which don't have one explicitly set:

- We'll import many trains at once, and we don't want to fire off a
  bazillion of requests (one per path item).
- The import feature will only be used with an empty infrastructure
  anyways.
- At some point [1] will be merged and this whole logic can be
  dropped altogether since a missing secondary code will prioritize
  stations.

[1]: https://github.com/OpenRailAssociation/osrd/pull/10325

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: take DTO instead of MacroEditorState in createPacedAttributesFromTrainrun()

We'll need to use this function without a MacroEditorState for NGE
imports in the next commit.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: add support for importing NGE JSON files

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
editoast: take a list in macro node POST endpoint

This will allow us to import a whole bunch of macro nodes at once
in a single HTTP request, just like we do for train schedules.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: import macro nodes

Signed-off-by: Simon Ser <contact@emersion.fr>

leovalais commented on editoast/src/views/scenario/macro_nodes.rs:
```diff
@@ -215,19 +215,21 @@ async fn create(
+                let changesets: Result<Vec<_>, _> = data
+                    .into_iter()
+                    .map(|node| node.into_macro_node_changeset(scenario_id))
+                    .collect();
```
```suggestion
                let changesets = data
                    .into_iter()
                    .map(|node| node.into_macro_node_changeset(scenario_id))
                    .collect_vec();
```

there is no error path in `into_macro_node_changeset`, can you change it to return the changeste directly to simplify things a bit?

emersion replied:
Sure thing!

leovalais commented on editoast/src/views/scenario/macro_nodes.rs:
```diff
@@ -179,18 +179,18 @@ async fn list(
     Path((project_id, study_id, scenario_id)): Path<(i64, i64, i64)>,
-    Json(data): Json<MacroNodeForm>,
-) -> Result<Json<MacroNodeResponse>> {
+    Json(data): Json<Vec<MacroNodeForm>>,
+) -> Result<Json<Vec<MacroNodeResponse>>> {
```
Wdyt about wrapping the arrays of both the requets and the response in an object to avoid anonymous arrays? Like:

```json
{ "nodes": [...] }
```

(sometimes we do this, but that's not consistent)

emersion replied:
What is the motivation for this?

In general I'm not a fan of unnecessary wrapping. The wrapping would potentially make the endpoint more extensible. (Would be nice to pick a choice and stick to it in future APIs.)

leovalais replied:
> What is the motivation for this?

Mostly the extensibility of the schema. We don't change requests / responses in such way often, but such wrapping has been useful a few times in editoast before. Considering how much it costs to wrap the array, even if this endpoint is not impacted by such refactoring ever, it's probably fine. Naming the content of the array can also make the json easier to read.

> Would be nice to pick a choice and stick to it in future APIs

Agreed. When the case arises I ask around, it doesn't seem that wrapping arrays is shocking anyone. We'd need an homogeneity pass, whatever we decide, indeed.

emersion replied:
I've pushed a fixup to wrap lists in structs, let me know if it's to your taste!

leovalais reviewed (approved):
LGTM about editoast! (A couple non-blocking comments.)

emersion committed:
editoast: stop using Result in MacroNodeForm::into_macro_node_changeset

This function can't fail.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: ensure macro node position is an integer

The backend will reject floating point numbers.

Signed-off-by: Simon Ser <contact@emersion.fr>

kmer2016 reviewed (approved):
LGTM !

louisgreiner commented:
Something weird happens on my side when I test the nodes and trainruns labels.

On this file:
[reticulaire(2).json](https://github.com/user-attachments/files/20587876/reticulaire.2.json)

I set the label "eee" to node SBG and "aze" to all my trainruns. However, after the import in OSRD, the "aze" label is not set on my trainruns anymore, but on the node SBG... I could not find the cause

louisgreiner commented on front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:
```diff
@@ -282,12 +244,48 @@ const generateTrainrunProperties = async (
+};
+
+const createPacedAttributesFromTrainrun = (trainrun: TrainrunDto, dto: NetzgrafikDto) => {
+  if (trainrun.frequencyId === TRAIN_SCHEDULE_FREQUENCY_ID) {
```
I think we have an issue here : `TRAIN_SCHEDULE_FREQUENCY_ID` is something very specific to our case, and this concept (unique train, non-paced) does not exist elsewhere. It happens that on [NGE (with default frequencies)](https://schweizerischebundesbahnen.github.io/netzgrafik-editor-frontend/), the `TRAIN_SCHEDULE_FREQUENCY_ID` actually refers to frequency 20min.

However, this workflow works when one exports a reticular from OSRD/NGE and imports it in OSRD/NGE (same "frequency concepts" on both sides). But this doesn't work anymore when a user exports from NGE (default frequencies let's say) and imports in OSRD/NGE : we cannot match the frequency ids for that case.

I'm not sure to know what should be the right solution. Maybe we'd need 2 different imports? Or just focus on "Actual NGE -> OSRD/NGE", and just create another custom frequency for that case?

louisgreiner replied:
I also realize that the fact that we match the frequencies with the ids in `getFrequencyFromFrequencyId()` makes it not resilient to different frequencies when we import a file with different frequencies.

With the current state of this PR, we are actually lucky that the frequencies ids of the imported file actually match our frequencies in OSRD/NGE

Reminder:
- id 0 : Frequency 15 (default NGE frequencies) / nothing (OSRD/NGE)
- id 1 : Frequency 20 (default NGE frequencies) / TrainSchedule (OSRD/NGE)
- id 2 : Frequency 30 (default NGE frequencies) / Frequency 30 (OSRD/NGE)
- id 3 : Frequency 60 (default NGE frequencies) / Frequency 60 (OSRD/NGE)
- id 4 : Frequency 120 (default NGE frequencies) / Frequency 120 (OSRD/NGE)
- id 5 : Frequency 120+ (default NGE frequencies) / Custom Frequency (OSRD/NGE)

And this can be also different when different Frequency set is imported. Maybe we should try to match rather on the `frequency` value, and dynamically create a custom Frequency when not match by our `DEFAULT_TRAINRUN_FREQUENCIES`?

I'm up for thinking on it if needed.

louisgreiner replied:
I thought on something like that, instead of using `createPacedAttributesFromTrainrun()` for the NGE import, use this function, which creates custom frequencies.

The below `currentFrequencies` should be initialized like this:
```ts
  const currentFrequencies = getDefaultTrainrunFrequencies(t);
```

```ts
export function matchOrCreateFrequency(
  trainrun: TrainrunDto,
  dto: NetzgrafikDto,
  currentFrequencies: TrainrunFrequency[]
) {
  const incomingFrequencyValue = dto.metadata.trainrunFrequencies.find(
    (freq) => freq.id === trainrun.frequencyId
  )?.frequency; // get the frequency actual value
  if (!incomingFrequencyValue) {
    throw new Error(`Frequency with ID ${trainrun.frequencyId} not found in metadata`);
  }

  // Try to find a matching frequency in the already existing frequencies
  let frequency: TrainrunFrequency | undefined;
  frequency = currentFrequencies.find((freq) => freq.frequency === incomingFrequencyValue);
  if (!frequency) {
    // if not, create a custom one and add it
    frequency = {
      id: currentFrequencies.length + 1,
      order: currentFrequencies.length + 1,
      frequency: incomingFrequencyValue,
      offset: 0,
      name: `Interval ${incomingFrequencyValue} min`, // todo: i18n
      shortName: `${incomingFrequencyValue}`,
      linePatternRef: '60',
    };
    currentFrequencies.push(frequency);
  }

  return {
    interval: new Duration({
      minutes: frequency.frequency,
    }).toISOString(),
    time_window: DEFAULT_TIME_WINDOW.toISOString(),
  };
}
```

emersion replied:
>I think we have an issue here : TRAIN_SCHEDULE_FREQUENCY_ID is something very specific to our case, and this concept (unique train, non-paced) does not exist elsewhere. It happens that on [NGE (with default frequencies)](https://schweizerischebundesbahnen.github.io/netzgrafik-editor-frontend/), the TRAIN_SCHEDULE_FREQUENCY_ID actually refers to frequency 20min.

That's a very good point!

What about the following: if the frequency is higher than 24h, then assume it's a single train schedule? (We could also check if the frequency is higher than the time window, but there is no way to select it when importing and I'm worried about data loss.)

> With the current state of this PR, we are actually lucky that the frequencies ids of the imported file actually match our frequencies in OSRD/NGE

I don't think so: `createPacedAttributesFromTrainrun()` already uses the DTO's frequencies' value in minutes and doesn't use `DEFAULT_TRAINRUN_FREQUENCIES`.

louisgreiner replied:
Good idea for the 24h+ frequencies as train schedules. I also fear that we could lose the 2 hours+ frequencies, if they exist, so I would not recommend (but maybe rather, if `frequency > 120`, then the time_window value is the same as the frequency value)

emersion replied:
I've pushed a commit to do that: "front: don't rely on TRAIN_SCHEDULE_FREQUENCY_ID in ngeToOsrd".

> if frequency > 120, then the time_window value is the same as the frequency value

I'm not sure this will work correctly: generally no train in a timetable has a frequency equal to the time window.

Maybe this should be a knob the user could set when importing trains, or even better a setting in NGE which gets exported in the DTO.

louisgreiner replied:
I tested and it seems it works fine with your commit

louisgreiner commented on front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:
```diff
@@ -683,4 +671,40 @@ const handleOperation = async ({
+
+  const trainSchedules: TrainSchedule[] = [];
+  const pacedTrains: PacedTrain[] = [];
+  for (const trainrun of dto.trainruns) {
+    const { path, labels, startDate, schedule } = generateTrainrunProperties(dto, trainrun);
```
Another point is that the import function is very strict, on the quality of the file imported. Here I try to import [this network graphic](https://schweizerischebundesbahnen.github.io/netzgrafik-editor-frontend/)

The current state of this feature expects a file without the potential errors of `getTrainrunSectionsByTrainrunId()`:
- not continuous trainrun
- missing departure section

which is very restrictive, either accepting or refusing a whole file : would it be better to allow to import as much trains as possible, and skip the errors?

```suggestion
  for (const trainrun of dto.trainruns) {
    if (!getTrainrunSectionsByTrainrunId(dto, trainrun.id)) continue;
    const { path, labels, startDate, schedule } = generateTrainrunProperties(dto, trainrun);
```

And allowing the main func to return null instead (or something similar):
```ts
  // if (orderedSections.length !== sections.length) {
  //   throw new Error('Trainrun is not continuous');
  // }
  if (orderedSections.length !== sections.length) {
    console.error(`Trainrun ${trainrunId} is not continuous`);
    return null;
  }
```

emersion replied:
I would prefer to be as strict as possible for now, and incrementally work our way towards handling more files by enhancing `ngeToOsrd`.

I do agree that we'll never be able to support e.g. disjoint trainruns, so a mechanism to report to the user which trains failed the import would be nicer. But I think informing the user is important here.

louisgreiner reviewed (commented):
Except the above points, the refactored code of the front-end part is so much clearer than before, it's amazing

emersion committed:
front: don't rely on TRAIN_SCHEDULE_FREQUENCY_ID in ngeToOsrd

This function will be used when importing foreign NGE files. In the
wild, NGE files will not use the same time frequencies as OSRD. In
particular, we can't rely on IDs used in the default OSRD time
categories.

Instead, only rely on the frequency value in minutes, and assume that
a frequency of 24 hours or more describes a single train.

Signed-off-by: Simon Ser <contact@emersion.fr>

flomonster reviewed (approved):
Editoast part LGTM.

louisgreiner commented on front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:
```diff
@@ -283,12 +244,47 @@ const generateTrainrunProperties = async (
+    return null;
+  }
+  return {
+    interval: interval.toISOString(),
+    time_window: DEFAULT_TIME_WINDOW.toISOString(),
```
I was a afraid of having `time_window > interval`, but after testing it seems that it is allowed and doesn't make any bug, nice

louisgreiner reviewed (approved):
Very nice !!

![image](https://github.com/user-attachments/assets/19a3510c-9f6d-4569-9927-53ce9c2f8009)
![image](https://github.com/user-attachments/assets/bbad79fb-a0c3-428d-b91c-c8148ffb80a1)

Besides the following bugs during import, it works very well, even with different frequency references.

- The trainrun labels are not displayed in NGE (but actually they are present in micro mode) (https://github.com/OpenRailAssociation/osrd/issues/12085)
- The first node (?) gets the previously deleted trainrun label (a `500` error happens after import)
`{"status":500,"type":"editoast:ModelError","context":{},"message":"an error occurred while querying the database: current transaction is aborted, commands ignored until end of transaction block"}`

I will open new bugs to deal with them, because it seems not to be related to this feature.
